[flang][OpenMP] Enable tiling

jsjodin opened the pull request:
This patch enables tiling in flang. It modifies the front-end to handle tiling as a nested construct if it is enlosed by an outer loop construct, or it can be stand-alone. In MLIR tiling is handled by changing the the omp.loop_nest op to be able to represent both collapse and tiling. In the MLIR->LLVM-IR the lowering of the LoopNestOp is enhanced to first do the tiling if present, then collapse. 

Notes/Details:
We are able to do tiling, and combine tiling with collapse. In the parse tree after canonicalization the tile loop construct becomes nested inside any outer loop construct.  The omp.loop_nest op has been enhanced to explicitly have a 'collapse' and 'tiles' attribute so that the transforms can be applied when lowering the code to LLVM-IR. This is a temporary solution since we are working on the full solution using canonical loops and loop transformation ops. The idea was to make a minimal, but still reasonable solution. Maybe there are some some ways to clean up this code. Please let me know if you have some ideas on how to do this better. One we are happy with the structure I can fix any typos/formatting and make this into a real PR if we decide to go ahead and merge this.

jsjodin committed:
Initial implementation of tiling.

jsjodin committed:
Fix tests and limit the nesting of construct to only tiling.

jsjodin committed:
Enable stand-alone tiling, but it gives a warning and converting to simd.

jsjodin committed:
Add minimal test, remove debug print.

jsjodin committed:
Fix formatting

jsjodin committed:
Fix formatting

ergawy commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -38,14 +38,21 @@ namespace lower {
   }
-  return 1;
+
+  collapseValue = collapseValue - numTileSizes;
+  int64_t result = collapseValue > numTileSizes ? collapseValue : numTileSizes;
+  return result;
```
Can you explain to me why we need this calculation? To see the effect of this, I tried replacing these few lines with simply `return collapseValue;` and ran all tests but no tests failed. So it seems this part is not tested. A test can also help explaining the purpose of the change.

jsjodin replied:
The general thing this computes is the number of loops that need to be considered in the source code. If you have collapse(4) on a loop nest with 2 loops that would be incorrect since we can max collapse 2 loops. However tiling creates new loops, so collapse(4) would theoretically be legal if tiling is done first e.g. tile(5,10) since that will result in 4 loops. This is not really testable though since collapse requires independent loops, which is only true for the 2 outer loops after tiling is done. There is a check for this, and an error message is given if the collapse value is larger than the number of loops that are tiled to prevent incorrect code. We could just use numTileSizes if that is present, but if collapse could handle dependent loops in the future the above calculation should be the correct one.

Meinersbur replied:
Support for collapsing grid- and tile-loops have only been added in OpenMP 6.0. Before that, there was a restriction:
![image](https://github.com/user-attachments/assets/75a23fac-3cbb-4004-87f4-ef0c20a49310)

The value of the collapse clause should not really be modified implicitly. Is there is check in FortranSemantics that ensures collapse is not larger than the number of grid loops?

jsjodin replied:
I added a check in resolve-directives.cpp (line 2009) that checks that the collapse is not larger than the number of tiled loops (I think this corresponds to the number of grid loops) and gives an error otherwise. Perhaps the error message could be improved.

ergawy commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -626,7 +634,16 @@ bool collectLoopRelatedInfo(
     collapseValue = evaluate::ToInt64(clause->v).value();
     found = true;
   }
+  std::int64_t sizesLengthValue = 0l;
+  if (auto *clause =
+          ClauseFinder::findUniqueClause<omp::clause::Sizes>(clauses)) {
+    sizesLengthValue = clause->v.size();
+    found = true;
+  }
 
+  collapseValue = collapseValue - sizesLengthValue;
+  collapseValue =
+      collapseValue < sizesLengthValue ? sizesLengthValue : collapseValue;
```
This as well does not seem to affect on tests added/modified by the PR.

Let me know if I missed something.

ergawy commented on flang/test/Lower/OpenMP/wsloop-tile.f90:
```diff
@@ -0,0 +1,30 @@
+  c=50
+  x=0
+
+  !CHECK: omp.loop_nest
+  !CHECK-SAME: tiles(2, 5, 10)
```
Can you add more details check lines for the loop nest structure?

Looking at the generated IR, we have 3 ops to represent the loop:
1. The `i` loop is an `omp.loop_nest` with a `tile` clause attached to it.
2. The `j` and `k` loops are `fir.do_loop` ops.

I kind of expected that the 3 loops will be collapsed into one `omp.loop_nest` op. But I am probably still missing the big picture here.

jsjodin replied:
It should be a single omp.loop_nest op, if not then it is a bug.

jsjodin replied:
I ran it on my machine and I'm getting this:
    omp.wsloop private(@_QFtripleloopEi_private_i32 %8#0 -> %arg1, @_QFtripleloopEj_private_i32 %10#0 -> %arg2, @_QFtripleloopEk_private_i32 %12#0 -> %arg3 : !fir.ref<i32>, !fir.ref<i32>, !fir.ref<i32>) {
      omp.loop_nest (%arg4, %arg5, %arg6) : i32 = (%c1_i32, %c1_i32_1, %c1_i32_3) to (%14, %15, %16) inclusive step (%c1_i32_0, %c1_i32_2, %c1_i32_4) tiles(2, 5, 10) {

jsjodin replied:
No problem, I added checks so that collapse(N) or tiles(1..N) makes sure that N is not greater than the number of loops in the omp.loop_nest op. It doesn't take into account the interaction between tiling and collapse, but we could add this also.

kparzysz commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -131,20 +133,48 @@ class CanonicalizationOfOmp {
-              std::get<std::optional<parser::OmpEndLoopDirective>>(x.t) =
-                  std::move(*endDir);
-              block.erase(nextIt);
+              std::get<std::optional<parser::OmpEndLoopDirective>>(
+                  loops.top()->t) = std::move(*endDir);
```
This should check if the actual directive matches the "begin loop" directive, e.g
```
!$omp do
!$omp tile ...
do ...
enddo
!$omp end do    ! should be paired with "do", not "tile"
```

jsjodin replied:
I looked into this and before this there was not checking for any mismatches, however in CheckDirectiveStructure.h is the code that makes sure that things are matched properly and errors out if not. I will take a look at this code and see if it needs improvement in any way. If there are no tests cases using tile in clause-validiy01/02 I can add that also.

jsjodin replied:
I went with the special handling.

kparzysz replied:
> Is the !$omp end tile optional? In that case we would need a special case for tile, since generally the mismatches are not handled here. If it is not optional then it should be handled later on in the checker.

Yes, for most of the loop directives the end-directive is optional.

kparzysz commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -131,20 +133,48 @@ class CanonicalizationOfOmp {
+              std::get<parser::OmpBeginLoopDirective>(innerOmpLoop->t)};
+          auto &innerDir{std::get<parser::OmpLoopDirective>(innerBeginDir.t)};
+          if (innerDir.v == llvm::omp::Directive::OMPD_tile) {
+            std::get<std::optional<
+                common::Indirection<parser::OpenMPLoopConstruct>>>(
+                loops.top()->t) = std::move(*innerOmpLoop);
```
Could you assign the lhs to a variable?  You're using the same expression again in the "loops.push" call.

jsjodin replied:
Done

kparzysz commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1890,39 +1910,124 @@ bool OmpAttributeVisitor::Pre(const parser::DoConstruct &x) {
+      std::get<std::optional<common::Indirection<parser::OpenMPLoopConstruct>>>(
+          x.t);
+  if (innerOptional.has_value()) {
+    CollectAssociatedLoopLevelsFromLoopConstruct(
+        innerOptional.value().value(), levels, clauses);
```
Doesn't the meaning of "level" change (by something like +1 or -1) when going to an inner construct?

jsjodin replied:
The level here refers to the number of loops in the source code that eventually get combined into a single omp.loop_nest op, so it will be the number of tiled loops or the collapse value that will determine the level.

kparzysz commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1890,39 +1910,124 @@ bool OmpAttributeVisitor::Pre(const parser::DoConstruct &x) {
+    const parser::OmpClauseList &x, llvm::SmallVector<std::int64_t> &levels,
+    llvm::SmallVector<const parser::OmpClause *> &clauses) {
+  for (const auto &clause : x.v) {
+    CollectAssociatedLoopLevelFromClauseValue<parser::OmpClause::Ordered>(
+        clause, levels, clauses);
+    CollectAssociatedLoopLevelFromClauseValue<parser::OmpClause::Collapse>(
+        clause, levels, clauses);
+    CollectAssociatedLoopLevelFromClauseSize<parser::OmpClause::Sizes>(
+        clause, levels, clauses);
```
It would probably be easier to read if these functions were inlined here.

kparzysz commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1890,39 +1910,124 @@ bool OmpAttributeVisitor::Pre(const parser::DoConstruct &x) {
+  std::int64_t tileLevel = 0;
+  for (auto [level, clause] : llvm::zip_equal(levels, clauses)) {
+    if (isSizesClause(clause)) {
+      tileLevel = level;
```
Shouldn't it be `tileLevel += level`?  You can stack `tile`s...

jsjodin replied:
That is unsupported right now. We don't have a way of representing nested tiling in MLIR, but will be possible when the loop transformation ops are added.  I can add a check and report an error for now.

Meinersbur commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -131,20 +133,48 @@ class CanonicalizationOfOmp {
       if (GetConstructIf<parser::CompilerDirective>(*nextIt))
         continue;
+      // Keep track of the loops to handle the end loop directives
+      std::stack<parser::OpenMPLoopConstruct *> loops;
```
llvm::SmallVector would be preferred unless you need to store more than 4294967295 elements

https://llvm.org/docs/ProgrammersManual.html#vector

[std::stack](https://en.cppreference.com/w/cpp/container/stack.html) is just a wrapper around the actual container, `std::deque` by default, but std::vector is possible as well (a stack only allows adding/removing elements from one end anyway). Unfortuntely, [std::deque](https://devblogs.microsoft.com/oldnewthing/20230810-00/?p=108587) is a quite bulky container, e.g. 4096 bytes object size (!) with libc++.

jsjodin committed:
Fix test.

jsjodin committed:
Add more mlir tests. Set collapse value when lowering from SCF to OpenMP.

jsjodin committed:
Use llvm::SmallVector instead of std::stack

jsjodin committed:
Improve test a bit to make sure IVs are used as expected.

jsjodin commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -112,15 +112,17 @@ class CanonicalizationOfOmp {
+    ///    OmpBeginLoopDirective t-> OmpLoopDirective t-> Tile v-> OMP_tile]
     //   ExecutableConstruct -> DoConstruct
+    //   [ExecutableConstruct -> OmpEndLoopDirective]
     //   ExecutableConstruct -> OmpEndLoopDirective (if available)
```
That is the optional OmpEndLoopDirective associated with the optional BeginLoopDirective for the tile construct, so it is meant to be there. I will update the "After rewriting" comment to hopefully clarify what is going on.

jsjodin replied:
Done updating comments.

jsjodin committed:
Fix comments to clarify canonicalization.

jsjodin committed:
Special handling of tile directive when dealing with start end end loop directives.

jsjodin committed:
Inline functions.

jsjodin commented:
Making this into a regular PR and adding more reviewers since we are getting more into the details.

ergawy commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -5024,8 +5024,9 @@ struct OpenMPBlockConstruct {
-      : t({std::move(a), std::nullopt, std::nullopt}) {}
+      : t({std::move(a), std::nullopt, std::nullopt, std::nullopt}) {}
   std::tuple<OmpBeginLoopDirective, std::optional<DoConstruct>,
+      std::optional<common::Indirection<OpenMPLoopConstruct>>,
```
nit: can you add a comment here explaining that this is used mainly for tiling constructs?

tblah replied:
Does this strictly only support nesting of two loop constructs or do you plan for it to work like a linked list: with each loop construct pointing to the next inner-most construct?

Stylie777 replied:
> Allowing multiple nested ops would be the final step, at that point we would need to use the CLIs, canonical loops and loop transformation ops.

As part of my work for #110008 I have implemented nested ops for `tile` and `unroll` in the semantics checks (PR coming soon, just doing finishing touches). It does work slightly differently in that `std::optional<DoConstruct>` has become `std::optional<std::variant<DoConstruct, common::Indirection<OpenMPLoopConstruct>>>` in the Parse Tree. 

This approach stores either one or the other, and then Flang can determine which is being used. I personally think this is a better approach as then you are not opening the possibility for there being a `DoConstruct` and `OpenMPLoopConstruct`. It works in a similar way where it will link together the LoopConstructs until you reach a DoConstruct.

Its currently a standalone change, as this is not merged yet, but it would be good to agree on an approach so there is minimal changes made when one of these is merged.

Stylie777 replied:
Just for reference, my PR for the Semantics covering `unroll` and `tile` are here: https://github.com/llvm/llvm-project/pull/145917

There does seem to be some crossover with this patch, so some rebasing will be needed depending on which is merged first.

jsjodin replied:
I rebased this PR now and things seem to be working fine.

ergawy commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2131,6 +2170,42 @@ genLoopOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
   return loopOp;
 }
 
+static mlir::omp::LoopOp genTiledLoopOp(lower::AbstractConverter &converter,
```
Can we reuse `genLoopOp` instead of adding a new function?

ergawy reviewed (commented):
Took another look this time trying to understand the approach on a higher level. I have one concern/question. It seems like we are "hiding" the `tile` directive too early in the pipeline (i.e. during PFT to the `omp` dialect); both for the combined and the standalone cases. I would have expected to model `tile` as a separate op in the dialect and handle mapping combined and standalone versions later in the compiler. In the standalone case, for instance, we conflate the `loop` and `tile` directives in one op which overloads `omp.loop` with more semantics than it is intended to carry.

To me, the `tile` is similar to `simd` in the sense that it is a loop wrapper op that controls division of the loop nests' iterations. So I would have expected both to have similar modelling (i.e. a separate op that impelements the `LoopWrapperInterface`).

You probably have good reason not to do what I suggested so just asking for my understanding.

jsjodin commented on llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:
```diff
@@ -482,6 +484,28 @@ bool ConstructDecompositionT<C, H>::applyClause(
+      last.clauses.push_back(node);
+      return true;
+    } else {
+      //      llvm::errs() << "** OVERRIDING isAllowedClauseForDirective **\n";
```
@kparzysz I'm not sure if there is a way to override this in a cleaner way? Is it easy to introduce a new internal clause (e..g tile_sizes) and instead of just moving the sizes clause from tile to the outer construct, we instead translate it into the internal clause? I'm not sure if there are surrounding data structures that will make this hard, like location information and such?

jsjodin replied:
Yes, since we are moving the sizes clause to the parent directive from tile to whatever happens to enclose it. That's why I was thinking about adding a new internal clause to specify tiling (just like collapse is specified) in the case where the tiling is not stand alone. This will no longer be needed though once @Meinersbur's patch is merged and we would use a separate op for tiling instead.

jsjodin replied:
Instead of trying to pass information around, the parse tree is available when collecting the information to create the loop nest ops, so that's the solution I went with. We no longer try to move the sizes clause to the parent loop construct.

jsjodin committed:
Remove debug code.

jsjodin committed:
Reuse loop op lowering, add comment.

jsjodin committed:
Fix formatting.

jsjodin committed:
Remove curly braces.

jsjodin commented:
> > This is the next step that @Meinersbur is working on, but we wanted to make tiling available sooner.
> 
> Are you against a deadline for the tiling support? @Meinersbur's patch is already up for review and looked high quality to me. I don't think it is that far away from being approved. Maybe the tiling support should be based upon that patch.

Yes, there is a bit of a deadline, however @Meinersbur 's patch is close to being approved it is probably best to get that in first, and then rework this patch.

ergawy commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -112,15 +111,19 @@ class CanonicalizationOfOmp {
     // Original:
-    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
-    //     OmpBeginLoopDirective
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct t->
```
I found the implementation of this function quite complex to understand tbh. Specially with the 2 nested `while` loops (the old one and the one introduced by the PR).

While trying to understand what is going on, I tried to simpify it by flatterning the logic a little bit:
```c++
  void RewriteOpenMPLoopConstruct(parser::OpenMPLoopConstruct &x,
      parser::Block &block, parser::Block::iterator it) {
    // Check the sequence of DoConstruct and OmpEndLoopDirective
    // in the same iteration
    //
    // Original:
    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct t->
    //     OmpBeginLoopDirective t-> OmpLoopDirective
    //   [ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct u->
    //     OmpBeginLoopDirective t-> OmpLoopDirective t-> Tile v-> OMP_tile]
    //   ExecutableConstruct -> DoConstruct
    //   [ExecutableConstruct -> OmpEndLoopDirective] (note: tile)
    //   ExecutableConstruct -> OmpEndLoopDirective (if available)
    //
    // After rewriting:
    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct t->
    //     [OpenMPLoopConstruct t -> OmpBeginLoopDirective -> OmpLoopDirective
    //                               OmpEndLoopDirective] (note: tile)
    //     OmpBeginLoopDirective t -> OmpLoopDirective -> DoConstruct
    //     OmpEndLoopDirective (if available)
    parser::Block::iterator nextIt;
    auto &beginDir{std::get<parser::OmpBeginLoopDirective>(x.t)};
    auto &dir{std::get<parser::OmpLoopDirective>(beginDir.t)};

    // Ignore compiler directives.
    nextIt = std::find_if_not(++it, block.end(), [this](auto &y) {
      return GetConstructIf<parser::CompilerDirective>(y);
    });

    if (nextIt == block.end()) {
      return;
    }

    auto &optionalInnerTile = std::get<
        std::optional<common::Indirection<parser::OpenMPLoopConstruct>>>(x.t);

    if (auto *innerConstruct{
            GetConstructIf<parser::OpenMPConstruct>(*nextIt)}) {
      if (auto *innerOmpLoop{
              std::get_if<parser::OpenMPLoopConstruct>(&innerConstruct->u)}) {
        auto &innerBeginDir{
            std::get<parser::OmpBeginLoopDirective>(innerOmpLoop->t)};
        auto &innerDir{std::get<parser::OmpLoopDirective>(innerBeginDir.t)};
        if (innerDir.v == llvm::omp::Directive::OMPD_tile) {
          optionalInnerTile = std::move(*innerOmpLoop);
          // Retrieveing the address so that DoConstruct or inner loop can be
          // set later.
          nextIt = block.erase(nextIt);
        }
      }
    }

    if (auto *doCons{GetConstructIf<parser::DoConstruct>(*nextIt)}) {
      if (doCons->GetLoopControl()) {
        parser::OpenMPLoopConstruct *innermostLoop =
            optionalInnerTile ? &optionalInnerTile.value().value() : &x;
        std::get<std::optional<parser::DoConstruct>>(innermostLoop->t) =
            std::move(*doCons);
        nextIt = block.erase(nextIt);

      } else {
        messages_.Say(dir.source,
            "DO loop after the %s directive must have loop control"_err_en_US,
            parser::ToUpperCaseLetters(dir.source.ToString()));
      }
    } else {
      messages_.Say(dir.source,
          "A DO loop must follow the %s directive"_err_en_US,
          parser::ToUpperCaseLetters(dir.source.ToString()));
    }

    auto tryToMatchEndLoopDirective = [&, this](
                                          parser::OpenMPLoopConstruct *loop) {
      if (nextIt == block.end()) {
        return false;
      }

      if (auto *endDir{GetConstructIf<parser::OmpEndLoopDirective>(*nextIt)}) {
        auto &endOmpDirective{std::get<parser::OmpLoopDirective>(endDir->t)};
        auto &loopBegin{std::get<parser::OmpBeginLoopDirective>(loop->t)};
        auto &loopDir{std::get<parser::OmpLoopDirective>(loopBegin.t)};

        // If the directive is a tile we try to match the corresponding
        // end tile if it exsists. If it is not a tile directive we
        // always assign the end loop directive and fall back on the
        // existing directive structure checks.
        if (loopDir.v != llvm::omp::Directive::OMPD_tile ||
            loopDir.v == endOmpDirective.v) {
          std::get<std::optional<parser::OmpEndLoopDirective>>(loop->t) =
              std::move(*endDir);
          nextIt = block.erase(nextIt);
        }

        return true;
      }

      // If there is a mismatch bail out.
      return false;
    };

    if (optionalInnerTile) {
      if (tryToMatchEndLoopDirective(&optionalInnerTile.value().value())) {
        tryToMatchEndLoopDirective(&x);
      }
    } else {
      tryToMatchEndLoopDirective(&x);
    }
  }
```

This is a, hopefully, more simplified version that matches the original logic. Feel free to ignore if I got the logic wrong.

jsjodin committed:
Avoid attaching the sizes clause to the parent construct, instead find the tile
sizes through the parse tree when getting the information needed to create the
loop nest ops.

jsjodin committed:
Fix formatting

Stylie777 commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -131,20 +134,59 @@ class CanonicalizationOfOmp {
+      // Keep track of the loops to handle the end loop directives
+      llvm::SmallVector<parser::OpenMPLoopConstruct *> loops;
+      loops.push_back(&x);
+      if (auto *innerConstruct{
```
```suggestion
      if (auto *innerConstruct{GetOmpIf<parser::OpenMPLoopConstruct>(*nextIt)}) {
```

We can simplify here and get the OpenMPLoopConstruct directly.

Stylie777 commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -131,20 +134,59 @@ class CanonicalizationOfOmp {
       // Ignore compiler directives.
       if (GetConstructIf<parser::CompilerDirective>(*nextIt))
         continue;
+      // Keep track of the loops to handle the end loop directives
```
Can you show how the Parse Tree looks with these changes, and include tests for this?

I am currently working on something similar for `tile` and `unroll` as they can both be nested within an OMP Loop Construct as part of #110008 so the semantics allow for the nesting.

I am not working on lowering, only semantics support.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2992,18 +2991,60 @@ convertOmpLoopNest(Operation &opInst, llvm::IRBuilderBase &builder,
 
-  // Collapse loops. Store the insertion point because LoopInfos may get
-  // invalidated.
+  //  llvm::OpenMPIRBuilder::InsertPointTy afterIP = builder.saveIP();
```
Was keeping this unintentional?

jsjodin replied:
Unintentional, thanks for spotting it.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2992,18 +2991,60 @@ convertOmpLoopNest(Operation &opInst, llvm::IRBuilderBase &builder,
+      TileSizes.push_back(TileVal);
+    }
+
+    std::vector<llvm::CanonicalLoopInfo *> NewLoops =
```
nit: SmallVector

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2992,18 +2991,60 @@ convertOmpLoopNest(Operation &opInst, llvm::IRBuilderBase &builder,
+
+    // Collapse loops. Store the insertion point because LoopInfos may get
+    // invalidated.
+    auto AfterBB = NewLoops.front()->getAfter();
+    auto AfterAfterBB = AfterBB->getSingleSuccessor();
```
nit (feel free to ignore): personally I find it a bit error-prone when `auto` hides that something is a pointer.
```suggestion
    auto *AfterBB = NewLoops.front()->getAfter();
    auto *AfterAfterBB = AfterBB->getSingleSuccessor();
```

Alternatively, spelling out `auto` would be fine here too.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2992,18 +2991,60 @@ convertOmpLoopNest(Operation &opInst, llvm::IRBuilderBase &builder,
+    std::vector<llvm::CanonicalLoopInfo *> NewLoops =
+        ompBuilder->tileLoops(ompLoc.DL, loopInfos, TileSizes);
+
+    // Collapse loops. Store the insertion point because LoopInfos may get
```
I don't understand in what sense this is collapsing the loops

jsjodin replied:
Done. Stray comment, I updated it to make sense.

jsjodin committed:
Fix unparse and add a test for nested loop constructs.

jsjodin committed:
Use more convenient function to get OpenMPLoopConstruct. Fix comments.

jsjodin committed:
Fix formatting.

tblah reviewed (commented):
I'm not sure the semantic and op verification checks are always right. See llvm-test-suite/Fortran/gfortran/regression/gomp/tile-inner-loops-*.f90. Especially tile-inner-loops-2.f90 where an mlir op verification error gets past semantics.

Stylie777 commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -131,20 +134,55 @@ class CanonicalizationOfOmp {
           nextIt = block.erase(nextIt);
           // try to match OmpEndLoopDirective
-          if (nextIt != block.end()) {
+          while (nextIt != block.end() && !loops.empty()) {
```
I have working on a similar thing for both `tile` and `unroll` in #145917, but this takes a different approach to finding the EndDirective and nested loops. I think we should agree on a unified approach over both PR's to reduce the need for re-work once one is merged.

I am not sure the `loops` variable is needed, this can be done just using `block` and recursion (the approach I took in my PR)

Stylie777 replied:
Will resolve as #145917 is merged.

Meinersbur commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -659,7 +756,6 @@ bool collectLoopRelatedInfo(
   } while (collapseValue > 0);
 
   convertLoopBounds(converter, currentLocation, result, loopVarTypeSize);
-
```
[nit] unrelated change

Meinersbur commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2992,18 +2991,59 @@ convertOmpLoopNest(Operation &opInst, llvm::IRBuilderBase &builder,
+    SmallVector<llvm::CanonicalLoopInfo *> collapseLoopInfos(
+        loopInfos.begin(), loopInfos.begin() + (numCollapse));
+
+    auto newLoopInfo =
+        ompBuilder->collapseLoops(ompLoc.DL, collapseLoopInfos, {});
```
Previously this was called withing the OpenMPLoopInfoStackFrame callback. Does it make a difference?

(Doing it outside of it seems more logical to me, but I assume there was a reason)

jsjodin committed:
Fix merge problems related to the different representations used for nested
loop constructs.

jsjodin committed:
Fix bugs introduced when merging.

jsjodin committed:
Move include

Meinersbur commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1974,44 +1986,124 @@ bool OmpAttributeVisitor::Pre(const parser::DoConstruct &x) {
+      context_.Say(clause->source,
+          "The value of the parameter in the COLLAPSE clause must"
+          " not be larger than the number of the number of tiled loops"
+          " because collapse relies on independent loop iterations."_err_en_US);
```
`collapse` does not require independent loops, non-rectangular loops are permitted (but which we currently do not support)

jsjodin replied:
Updating error message.

Meinersbur commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -816,7 +816,20 @@ class OmpAttributeVisitor : DirectiveAttributeVisitor<llvm::omp::Directive> {
 private:
+  std::int64_t SetAssociatedMaxClause(llvm::SmallVector<std::int64_t> &,
+      llvm::SmallVector<const parser::OmpClause *> &);
+  std::int64_t GetAssociatedLoopLevelFromLoopConstruct(
+      const parser::OpenMPLoopConstruct &);
   std::int64_t GetAssociatedLoopLevelFromClauses(const parser::OmpClauseList &);
+  void CollectAssociatedLoopLevelsFromLoopConstruct(
+      const parser::OpenMPLoopConstruct &, llvm::SmallVector<std::int64_t> &,
+      llvm::SmallVector<const parser::OmpClause *> &);
+  void CollectAssociatedLoopLevelsFromInnerLoopContruct(
+      const parser::OpenMPLoopConstruct &, llvm::SmallVector<std::int64_t> &,
+      llvm::SmallVector<const parser::OmpClause *> &);
+  void CollectAssociatedLoopLevelsFromClauses(const parser::OmpClauseList &,
+      llvm::SmallVector<std::int64_t> &,
+      llvm::SmallVector<const parser::OmpClause *> &);
```
If these do not require `OmpAttributeVisitor`, consider making them standalone static functions.

jsjodin replied:
We're using some members in the class currently. I think I tried making them static functions initially, but it didn't work out very well.

jsjodin committed:
Remove unused code. Currently the canonicalize-omp can only handle a single
nested loop construct, which is what we prefer.

jsjodin commented on mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:
```diff
@@ -317,6 +317,38 @@ class OpenMP_DeviceClauseSkip<
+// V5.2: [xx.x] `sizes` clause
+//===----------------------------------------------------------------------===//
+
+class OpenMP_TileSizesClauseSkip<
+    bit traits = false, bit arguments = false, bit assemblyFormat = false,
+    bit description = false, bit extraClassDeclaration = false
+  > : OpenMP_Clause<traits, arguments, assemblyFormat, description,
+                    extraClassDeclaration> {
+  let arguments = (ins
+      OptionalAttr<DenseI64ArrayAttr>:$tile_sizes
+  );
+}
+
+def OpenMP_TileSizesClause : OpenMP_TileSizesClauseSkip<>;
```
I wanted to limit this to constant values initially. I can add the OpenMP_SizesClauseSkip class in a separate PR and retire this code then.

jsjodin committed:
Address review comments.

jsjodin committed:
Undo unrelated change.

Meinersbur reviewed (approved):
LGTM. Please also wait for an approval from @skatrak

Meinersbur reviewed (changes requested):
Sorry, had another look and found a bigger issue.

Meinersbur reviewed (changes requested):
Sorry, had another look and found a bigger issue.

jsjodin commented on llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:
```diff
@@ -482,6 +484,24 @@ bool ConstructDecompositionT<C, H>::applyClause(
 
   return false;
 }
+// FIXME(JAN): Do the correct thing, but for now we'll do the same as collapse
```
I tried using this, but since the size clause is not allowed in the position I was planning to use it, it will report an error.

jsjodin commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -817,7 +817,24 @@ class OmpAttributeVisitor : DirectiveAttributeVisitor<llvm::omp::Directive> {
+      llvm::SmallVector<const parser::OmpClause *> &);
+  std::int64_t GetNumAffectedLoopsFromLoopConstruct(
+      const parser::OpenMPLoopConstruct &);
+  std::int64_t GetNumAffectedLoopsFromClauses(const parser::OmpClauseList &);
```
I will remove this function.

jsjodin commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -3449,13 +3500,9 @@ static void genOMPDispatch(lower::AbstractConverter &converter,
-      TODO(loc, "Unhandled loop directive (" +
-                    llvm::omp::getOpenMPDirectiveName(dir, version) + ")");
+  case llvm::omp::Directive::OMPD_tile:
+    newOp = genLoopOp(converter, symTable, semaCtx, eval, loc, queue, item);
```
I may not be remembering this correctly though, it was some time back. I believe the LoopOp is actually added somewhere earlier, but I have to look into it.

Meinersbur replied:
> In the GenericLoopConversionPass there is a function rewriteStandaloneLoop that converts the loop to a simdloop. 

I think the semantics are different. Independent of how it is executed (parallel, simd, sequential, ...), the loop is duplicated with a copy nested in the other. This logically results in $n^2$ executions, where $n$ is the tripcount of the original loop. 

I was experimenting with it, and could observe those additional executions.

jsjodin replied:
I looked a bit more closely and this will require some other things to be fixed, so I will leave it for a different PR and revert back to not handling the stand-alone tiling.

jsjodin committed:
Remove stand-alone tiling.

jsjodin committed:
Revert unused changes.

jsjodin committed:
Don't do codegen for tiling if it is an inner construct.

Meinersbur commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -3917,8 +3968,8 @@ static void genOMP(lower::AbstractConverter &converter, lower::SymMap &symTable,
       case llvm::omp::Directive::OMPD_tile:
-        // Emit the omp.loop_nest with annotation for tiling
-        genOMP(converter, symTable, semaCtx, eval, ompNestedLoopCons->value());
+        // Skip OMPD_tile since the tile sizes will be retrieved when
+        // generating the omp.looop_nest op.
```
```suggestion
        // Skip OMPD_tile since the tile sizes will be retrieved when
        // generating the omp.loop_nest op.
```

The entire code in `if (optLoopCons.has_value()) {...}` is only intended to raise a `TODO` error?

jsjodin replied:
Yes, for right now, but when the lowering is replaced with loop transformations and canonical loops it won't or if other people add nested constructs for some other use they can expand this code.

Meinersbur commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -1660,6 +1687,30 @@ genLoopNestClauses(lower::AbstractConverter &converter,
+        int64_t sizeValue = evaluate::ToInt64(size).value();
+        sizeValues.push_back(sizeValue);
+      }
+      clauseOps.tileSizes = sizeValues;
+    }
+  }
+
+  llvm::SmallVector<int64_t> sizeValues;
+  auto *ompCons{eval.getIf<parser::OpenMPConstruct>()};
+  collectTileSizesFromOpenMPConstruct(ompCons, sizeValues, semaCtx);
+  if (sizeValues.size() > 0)
+    clauseOps.tileSizes = sizeValues;
```
Are these two redundant means to fetch the tile size?

jsjodin replied:
The stand-alone case is no longer needed in this PR, so I can delete it.

jsjodin committed:
Remove unused stand-alone tiling code. Fix typo.

jsjodin committed:
Remove unused code.

skatrak reviewed (commented):
Thank you Jan for all the work on this and excuse how long it's taken me to take a good look at it. I have a couple of questions, mainly about how the MLIR representation changes work, and some comments.

jsjodin commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -418,6 +420,28 @@ static void processHostEvalClauses(lower::AbstractConverter &converter,
+                  middleClauseList =
+                      &std::get<parser::OmpClauseList>(innerBegin.t);
+                }
+              }
```
This was removed since we are only using constants for now. Later PRs will have to update this.

jsjodin commented on mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:
```diff
@@ -2956,17 +2961,46 @@ ParseResult LoopNestOp::parse(OpAsmParser &parser, OperationState &result) {
+    return success();
+  };
+
+  if (!parser.parseOptionalKeyword("tiles") &&
```
It was a combination of tile and sizes, It doesn't really matter which one we pick imo.

jsjodin committed:
Address review comments

jsjodin committed:
Move include line

jsjodin committed:
Move include to correct location hopefully.

skatrak commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -271,12 +271,25 @@ bool ClauseProcessor::processCancelDirectiveName(
-bool ClauseProcessor::processCollapse(
+bool ClauseProcessor::processLoopNests(
     mlir::Location currentLocation, lower::pft::Evaluation &eval,
     mlir::omp::LoopRelatedClauseOps &result,
     llvm::SmallVectorImpl<const semantics::Symbol *> &iv) const {
-  return collectLoopRelatedInfo(converter, currentLocation, eval, clauses,
-                                result, iv);
+  int64_t numCollapse = collectLoopRelatedInfo(converter, currentLocation, eval,
+                                               clauses, result, iv);
+  return numCollapse > 1;
+}
```
I think we shouldn't introduce this function and instead replace uses with `processCollapse`. By calling this instead of `processCollapse` in `processHostEvalClauses`, the result is that we discard collapse information on the host, so host fallback will behave as if the clause wasn't there. For example:
```f90
subroutine test()
  implicit none
  integer :: i, j

  !$omp target teams distribute parallel do collapse(2)
  do i=1,10
    do j=1,20
      call foo(i, j)
    end do
  end do
end subroutine
```
Will result in the following MLIR for the `omp.loop_nest` for host and device:
```mlir
// Host
// flang -fc1 -emit-hlfir -fopenmp test.f90 -o - | grep omp.loop_nest
omp.loop_nest (%arg10, %arg11) : i32 = (%arg0, %arg1) to (%arg2, %arg3) inclusive step (%arg4, %arg5) { ... }

// Device
// flang -fc1 -emit-hlfir -fopenmp -fopenmp-is-target-device test.f90 -o - | grep omp.loop_nest
omp.loop_nest (%arg4, %arg5) : i32 = (%c1_i32, %c1_i32_1) to (%c10_i32, %c20_i32) inclusive step (%c1_i32_0, %c1_i32_2) collapse(2) { ... }
```

jsjodin replied:
I removed the new function and updated processCollapse.

skatrak reviewed (commented):
Thank you for the clarifications and updates Jan, I have just one blocking comment and some minor nits, but overall this looks almost ready to me.

Currently, there is a comment in `initTargetRuntimeAttrs` stating that it calculates the trip count by multiplying the trip counts of every collapsed canonical loop. I think that comment should be updated to basically just say that it's calculated by multiplying the trip counts of all nested loops represented by the op. This is just so it doesn't become confusing after the changes to `omp.loop_nest`.

jsjodin committed:
Make sure collapse is progagated correctly on the host side.

jsjodin committed:
Address more review comments.

skatrak reviewed (approved):
Thank you Jan again for all the work on this. LGTM, just some non-blocking nits left from me. Feel free to merge when everyone else is happy.

jsjodin commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -3940,8 +3942,8 @@ static void genOMP(lower::AbstractConverter &converter, lower::SymMap &symTable,
-        // Emit the omp.loop_nest with annotation for tiling
-        genOMP(converter, symTable, semaCtx, eval, ompNestedLoopCons->value());
+        // Skip OMPD_tile since the tile sizes will be retrieved when
+        // generating the omp.loop_nest op.
```
In that case it will not be a nested construct and will hit the TODO on line 3476.

jsjodin committed:
Address review comments

kparzysz reviewed (approved):
LGTM.  Thanks!

jsjodin commented:
> Hi Jan, you may want to document omp tile support in flang.

I was going to wait documenting this until the next couple of PRs are merged since we're limited to constants only and can't handle the stand-alone tiling case. Once those are done it is pretty easy to document without having to go into a lot of details of what doesn't work.

jsjodin committed:
Use shared helper function to find the sizes clause.

jsjodin commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -595,9 +660,43 @@ bool collectLoopRelatedInfo(
+        }
+      }
+    }
   }
```
I created a helper function with a callback so I didn't have to get the context in the other use.

jsjodin commented:
> @jsjodin looks like this patch breaks handling of collapse clause in target-based directives. E.g. last test in test/Lower/OpenMP/distribute-parallel-do-simd.f90 shall emit collapse attribute, but it is not.

I will work on a fix.
